core/trino-main/src/main/java/io/trino/sql/ir/optimizer/rule/SimplifyContinuousInValues.java

```diff
@@ -0,0 +1,127 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.ir.optimizer.rule;
+
+import io.trino.Session;
+import io.trino.spi.type.DecimalType;
+import io.trino.spi.type.Type;
+import io.trino.sql.ir.Between;
+import io.trino.sql.ir.Constant;
+import io.trino.sql.ir.Expression;
+import io.trino.sql.ir.In;
+import io.trino.sql.ir.IsNull;
+import io.trino.sql.ir.optimizer.IrOptimizerRule;
+import io.trino.sql.planner.Symbol;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.DateType.DATE;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.SmallintType.SMALLINT;
+import static io.trino.spi.type.TimeType.TIME_MICROS;
+import static io.trino.spi.type.TimeType.TIME_MILLIS;
+import static io.trino.spi.type.TimeType.TIME_NANOS;
+import static io.trino.spi.type.TimeType.TIME_PICOS;
+import static io.trino.spi.type.TimeType.TIME_SECONDS;
+import static io.trino.spi.type.TimestampType.TIMESTAMP_MICROS;
+import static io.trino.spi.type.TimestampType.TIMESTAMP_MILLIS;
+import static io.trino.spi.type.TimestampType.TIMESTAMP_SECONDS;
+import static io.trino.spi.type.TinyintType.TINYINT;
+import static io.trino.sql.ir.IrUtils.or;
+
+/**
+ * Simplify IN expression with continuous range of constant test values into a BETWEEN expression. E.g,
+ * <ul>
+ *     <li>{@code $in(x, [1, 2, 3, 4]) -> $between(x, 1, 4)}
+ * </ul>
+ */
+public class SimplifyContinuousInValues
+        implements IrOptimizerRule
+{
+    @Override
+    public Optional<Expression> apply(Expression expression, Session session, Map<Symbol, Expression> bindings)
+    {
+        if (!(expression instanceof In in)) {
+            return Optional.empty();
+        }
+
+        return simplifyContinuousInValues(in);
+    }
+
+    public static Optional<Expression> simplifyContinuousInValues(In in)
+    {
+        if (in.valueList().size() < 2) {
+            return Optional.empty();
+        }
+
+        Type valueType = in.value().type();
+        if (!isIntegerArithmeticValidType(valueType)) {
+            return Optional.empty();
+        }
+
+        boolean nullMatch = false;
+        long nonNullsCount = 0;
+        long min = Long.MAX_VALUE;
+        long max = Long.MIN_VALUE;
+        for (Expression testExpression : in.valueList()) {
+            if (!(testExpression instanceof Constant constant)) {
+                return Optional.empty();
+            }
+            if (constant.value() == null) {
+                nullMatch = true;
+                continue;
+            }
+            long longConstant = ((Number) constant.value()).longValue();
```

Comment:
Cast to long directly. Converting the value via `Number` and `longValue()` can mask issues if the underlying value is not actually a long.